tensor2tensor/rl/restarter.py

```diff
@@ -0,0 +1,124 @@
+    target_local_step (int): Local step to train the model up to.
+
+  Attributes:
+    model_mode (str): See args.
+    checkpoint_dir (str): See args.
+    target_local_step (int): See args.
+    target_global_step (int): Calculated global step to train the model up to.
+    should_skip (bool): Whether training should be skipped because the number of
+        local steps already done is higher than the target. This happens during
+        restarts.
+    restarting (bool): Whether the current epoch of training has been
+        interrupted and is being restarted.
+  """
+
+  def __init__(self, model_mode, checkpoint_dir, target_local_step):
+    self.model_mode = model_mode
+    self.checkpoint_dir = checkpoint_dir
+    self.target_local_step = target_local_step
+    self.target_global_step = None
+    self.should_skip = False
+    self.restarting = False
+
+    self._counter_path = os.path.join(
+        checkpoint_dir, "{}_step_counter".format(model_mode)
+    )
+
+    self._global_step = self._get_global_step()
+    tf.logging.info(
+        "Will load %s checkpoint %d", self.model_mode, self._global_step
+    )
+
+    (self._local_step_at_start, global_step_at_start) = self._read_counters()
+
+    self.steps_to_go = target_local_step - self._local_step_at_start
+    if self.steps_to_go <= 0:
+      tf.logging.info(
+          "Skipping training %s, requested %d steps, already done %d",
+          self.model_mode, target_local_step, self._local_step_at_start
+      )
+      self.should_skip = True
+      return
+
+    if global_step_at_start != -1:
+      # Restart.
+      steps_done_this_epoch = self._global_step - global_step_at_start
+      self.steps_to_go -= steps_done_this_epoch
+      tf.logging.info(
+          "Restarting training %s, %d steps already done this epoch",
+          self.model_mode, steps_done_this_epoch
+      )
+      self.restarting = True
+
+    self.target_global_step = self._global_step + self.steps_to_go
+
+  @contextmanager
+  def training_loop(self):
+    """Context manager wrapping the training loop.
+
+    Takes care of updating the step counters.
+    """
+    if not self.restarting:
+      self._write_counters(self._local_step_at_start, self._global_step)
+
+    tf.logging.info(
+        "Training %s up to %d, %d to go", self.model_mode,
+        self.target_local_step, self.steps_to_go
+    )
+
+    yield
+
+    self._write_counters(self.target_local_step, -1)
+
+  def _get_global_step(self):
+    checkpoint = tf.train.latest_checkpoint(self.checkpoint_dir)
+    if checkpoint:
+      return int(checkpoint.split("-")[-1])
+    else:
+      return 0
+
+  def _read_counters(self):
+    try:
+      with open(self._counter_path, "r") as f:
+        return tuple(
+            int(counter) for counter in f.read().split(" ")
+        )
+    except FileNotFoundError:
+      return (0, -1)
+
+  def _write_counters(self, local_step, global_step):
+    with open(self._counter_path, "w") as f:
```

Comment:
Please access all files with tf file functions, so tf.gfile.Open(...) -- -otherwise it fails on other file systems.